迁移editbox

ftershock opened the pull request:
Re: https://github.com/cocos-creator/3d-tasks/issues/1600

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -245,32 +269,28 @@ export class EditBoxComponent extends Component {
     get fontColor () {
-        return this._fontColor;
+        if (!this.textLabel){
+            return math.Color.WHITE.clone();
```
为什么需要 clone？

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -245,32 +269,28 @@ export class EditBoxComponent extends Component {
             if (renderComp) {
-                renderComp.color = this._fontColor;
+                renderComp.color = value;
+                renderComp.opacity = value.a;
```
这个组件有 opacity 属性？

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -331,28 +343,25 @@ export class EditBoxComponent extends Component {
     get placeholderFontColor () {
-        return this._placeholderFontColor;
+        if (!this._placeholderLabel) {
+            return math.Color.GRAY.clone();
```
同上

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -331,28 +343,25 @@ export class EditBoxComponent extends Component {
-            }
-            // this._placeholderLabel.node.opacity = this._placeholderFontColor.a;
+            comp.color = value;
+            comp.opacity = value!.a;
```
同上

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -365,20 +374,13 @@ export class EditBoxComponent extends Component {
-            return;
-        }
-
+    set maxLength (value: any) {
```
为什么是 any 类型

JoneLau replied:
为什么是 any 类型

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -476,14 +480,21 @@ export class EditBoxComponent extends Component {
+    @property({
+        type: float,
+    })
+    private _N$maxLength;
```
为什么声明了没有用到？

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -476,14 +480,21 @@ export class EditBoxComponent extends Component {
     public _background: SpriteComponent | null = null;
+
+    @property({
+        type: float,
```
为什么要定义 float 类型

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -648,173 +563,214 @@ export class EditBoxComponent extends Component {
-            if (this._placeholderLabel) {
-                this._placeholderLabel.node.active = displayText === '';
-            }
+        public _showLabels () {
```
为什么代码格式歪了

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -835,18 +811,63 @@ export class EditBoxComponent extends Component {
+        }
+    }
+
+    public _resizeChildNodes () {
```
全部都是 public 方法？

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -835,18 +811,63 @@ export class EditBoxComponent extends Component {
+        const textLabel = this._textLabel;
+        if (textLabel) {
+            textLabel.node.setContentSize(size.width - LEFT_PADDING, size.height);
+            textLabel.node.setPosition(offx + LEFT_PADDING, offy + size.height, textLabel.node.getPosition().z);
```
这里之前说过了，getPosition 方法需要返回一个 out 参数，如果 out 没给会去 new 一个 Vec3，你这里明显没有必要，用 position

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -835,18 +811,63 @@ export class EditBoxComponent extends Component {
+    public _resizeChildNodes () {
+        const textLabelNode = this._textLabel && this._textLabel.node;
+        if (textLabelNode) {
+            textLabelNode.setPosition(-this.node.width / 2, this.node.height / 2, textLabelNode.getPosition().z);
```
position

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -835,18 +811,63 @@ export class EditBoxComponent extends Component {
+        }
+        const placeholderLabelNode = this._placeholderLabel && this._placeholderLabel.node;
+        if (placeholderLabelNode) {
+            placeholderLabelNode.setPosition(-this.node.width / 2, this.node.height / 2, placeholderLabelNode.getPosition().z);
```
position

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -648,173 +563,214 @@ export class EditBoxComponent extends Component {
-            placeholderLabelNode.width = this.node.width;
-            placeholderLabelNode.height = this.node.height;
+        if (this._background) {
+            this._background.node.getComponent(UITransformComponent).anchorPoint = this.node.getComponent(UITransformComponent).anchorPoint;
```
如果能保证 ui 节点，直接从 node.uiTransfromComp 获取

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -835,18 +811,63 @@ export class EditBoxComponent extends Component {
+            return this._impl.isFocused();
+        }
+        else {
+            return false;
```
不用再 else 了，直接 return false

JoneLau commented on cocos/ui/components/editbox/tabIndexUtil.ts:
```diff
@@ -0,0 +1,38 @@
+// tslint:disable-next-line: class-name
+export class tabIndexUtil {
+    static _tabIndexList: any = [];
```
没法确定类型？

JoneLau commented on cocos/ui/components/editbox/tabIndexUtil.ts:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    public static next (editBoxImpl) {
+        let list = this._tabIndexList;
```
这几个为啥都用 let，没有需要被修改呀

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -89,11 +89,47 @@ export class EditBoxComponent extends Component {
+    }
 
+    set placeholderLabel (oldValue) {
+        if (this.placeholderLabel && this.placeholderLabel !== oldValue) {
```
意思是没有 placeholderLabel 我不能设置？还有为什么这里不用 _placeholderLabel

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -281,23 +300,18 @@ export class EditBoxComponent extends Component {
+        if (!this._placeholderLabel) {
+            return '';
+        }
+        return this._placeholderLabel.getComponent(LabelComponent)!.string;
```
这个 _placeholderLabel 难道不是 LabelComponent？

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -331,28 +342,24 @@ export class EditBoxComponent extends Component {
+        if (!this._placeholderLabel) {
+            return math.Color.GRAY;
+        }
+        return this.placeholderLabel!.color;
```
为什么要用 get 的 placeholderLabel

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -331,28 +342,24 @@ export class EditBoxComponent extends Component {
-                comp.color = this._placeholderFontColor;
-            }
-            // this._placeholderLabel.node.opacity = this._placeholderFontColor.a;
+            comp.color = value;
```
这个 comp 是不是多此一举了

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -476,14 +478,17 @@ export class EditBoxComponent extends Component {
     public _background: SpriteComponent | null = null;
+
+    @property
+    private _textLabel: LabelComponent | null = null;
```
可继承的属性用 protected

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -648,86 +557,104 @@ export class EditBoxComponent extends Component {
-            this._textLabel.node.active = false;
+        this._isLabelVisible = false;
+        if (this.textLabel) {
+            this.textLabel.node.active = false;
```
这里也是一个，一定要用 get 的 textLabel?

JoneLau commented on cocos/ui/components/editbox/tabIndexUtil.ts:
```diff
@@ -0,0 +1,39 @@
+export class tabIndexUtil {
+    static _tabIndexList: EditBoxImpl[] = [];
+
+    public static add (editBoxImpl) {
```
editBoxImpl 值类型也定义一下

JoneLau replied:
editBoxImpl 值类型也定义一下

JoneLau commented on cocos/ui/components/editbox/tabIndexUtil.ts:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    public static resort () {
+        this._tabIndexList.sort((a, b) => {
```
a，b 类型也没有吗

JoneLau commented on cocos/ui/components/editbox/tabIndexUtil.ts:
```diff
@@ -0,0 +1,39 @@
+import { EditBoxImpl } from './edit-box-impl';
+
+export class tabIndexUtil {
+    static _tabIndexList: EditBoxImpl[] = [];
```
public 还是 proteced 还是 private

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -245,31 +275,23 @@ export class EditBoxComponent extends Component {
-        }
-        if (this._impl) {
-            this._impl.setFontColor(this._fontColor);
+                this._textLabel.node.getComponent(UIRenderComponent).color = value;
```
this._textLabel  就是一个继承自 UIRenderComponent 的组件，它自身就有 color，多看看文档，很多地方都是再去 getComponent 一次，都是没有必要的

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -331,28 +345,23 @@ export class EditBoxComponent extends Component {
-                comp.color = this._placeholderFontColor;
-            }
-            // this._placeholderLabel.node.opacity = this._placeholderFontColor.a;
+            this._placeholderLabel.node.getComponent(UIRenderComponent).color = value;
```
同上

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -476,14 +480,17 @@ export class EditBoxComponent extends Component {
+    @property
+    protected _placeholderLabel: LabelComponent | null = null;
     @property
     private _returnType = KeyboardReturnType.DEFAULT;
```
这里所有的需要序列化 properties 都改成 protected 比较好

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -648,81 +559,91 @@ export class EditBoxComponent extends Component {
             let node = this.node.getChildByName('TEXT_LABEL');
             if (!node) {
-                node = new Node('TEXT_LABEL') as INode;
+                node = new cc.Node('TEXT_LABEL');
```
不要用 cc.

JoneLau commented on cocos/ui/components/editbox/edit-box-component.ts:
```diff
@@ -648,81 +559,91 @@ export class EditBoxComponent extends Component {
             let node = this.node.getChildByName('PLACEHOLDER_LABEL');
             if (!node) {
-                node = new Node('PLACEHOLDER_LABEL') as INode;
+                node = new cc.Node('PLACEHOLDER_LABEL');
```
同上

JoneLau commented on cocos/ui/components/editbox/edit-box-impl.ts:
```diff
@@ -30,20 +30,23 @@
  * @hidden
  */
 
+import { BitmapFont, UITransformComponent } from '../../../core';
```
有几处无用引用

ftershock committed:
迁移editbox


迁移editbox


迁移editbox


迁移editbox


迁移editbox


迁移editbox


迁移editbox


迁移editbox


迁移editbox


迁移editbox


迁移editbox
